### ISSUE TITLE
apps v2 `fly deploy` hangs on release_command

### ISSUE THREAD

**Issue by tvdfly** [superfly/flyctl#2036]
I ran `fly deploy` on an app that has a release_command. The release command succeeded, but `fly deploy` hung forever waiting for the machine to have the started state.

This was the output from fly deploy:

```
--> Pushing image done
image: [registry.fly.io/tvd-analytics:deployment-01GXDTF8DQM3NX1D369K1RCVB6](http://registry.fly.io/tvd-analytics:deployment-01GXDTF8DQM3NX1D369K1RCVB6)
image size: 132 MB
Running tvd-analytics release_command: /release_cmd.sh
  Waiting for 5683976f7d7e8e to have state: started
```

logs show that the release command finished successfully:

```
2023-04-07T12:19:02.661 app[5683976f7d7e8e] sea [info] Migrations successful!
2023-04-07T12:19:02.766 app[5683976f7d7e8e] sea [info] Starting clean up.
2023-04-07T12:19:03.767 app[5683976f7d7e8e] sea [info] [ 4.137571] reboot: Restarting system
2023-04-07T12:19:04.663 runner[5683976f7d7e8e] sea [info] machine restart policy set to 'no', not restarting
```

The release_command machine was successfully destroyed:

```
% fly m status 5683976f7d7e8e
Machine ID: 5683976f7d7e8e
Instance ID: 01GXDTFDX2SVFAW1V6YQPG5761
State: destroyed

VM
  ID            = 5683976f7d7e8e
  Instance ID   = 01GXDTFDX2SVFAW1V6YQPG5761
  State         = destroyed
  Image         = tvd-analytics:deployment-01GXDTF8DQM3NX1D369K1RCVB6
  Name          = restless-breeze-5622
  Private IP    =
  Region        = sea
  Process Group = fly_app_release_command
  CPU Kind      = shared
  vCPUs         = 2
  Memory        = 512
  Created       = 2023-04-07T12:18:50Z
  Updated       = 2023-04-07T12:19:04Z
  Command       = /release_cmd.sh

Event Logs
STATE           EVENT   SOURCE  TIMESTAMP                       INFO
destroyed       destroy flyd    2023-04-07T05:19:04.786-07:00
destroying      destroy flyd    2023-04-07T05:19:04.664-07:00
stopped         exit    flyd    2023-04-07T05:19:04.663-07:00   exit_code=0,oom_killed=false,requested_stop=false
started         start   flyd    2023-04-07T05:18:59.764-07:00
created         launch  user    2023-04-07T05:18:50.933-07:00
```

ctrl+c did not work to kill the `fly deploy` process. kill -15 didn't work either. kill -9 did work.

There's probably a couple things at play here that could be fixed: 1) a bug in flyctl's handling of responses from the wait endpoint or maybe a bug in the wait endpoint, and 2) signal handling in flyctl is not working as expected in some cases.

**Comment by matttpt** [comment]
I'm able to reproduce something like this, though it's under weird conditions:

* If `WaitForState` in `runReleaseCommand` runs after the release command has already finished, it gets stuck in an error loop because the wait endpoint returns "machine not found". I had to introduce a `time.Sleep` call in `runReleaseCommand` to get this to happen, but I wonder if an intermittent network issue would be enough to cause this in real life.
* After 120 seconds of the error loop, the context times out, and because of the same issue with `rehttp` that turned up in #1801, `flyctl` hangs and can't be killed with ctrl-C.

This might not be what's happening to you, but I think there are two actionable items anyway:

* Ship a more general solution to the `rehttp` issue than #1801
* Handle properly the case when a release command finishes before we hit the wait endpoint

**Comment by tvdfly** [comment]
Thanks @matttpt, that makes sense to me. Would you like to take on these fixes? I'd start with this one: `Handle properly the case when a release command finishes before we hit the wait endpoint` to help other users avoid this issue.

**Comment by rugwirobaker** [comment]
Is this solved now?

**Comment by matttpt** [comment]
Pavel addressed the other part of this in #2080 (I forgot to post that here!). I'm going to close with the hope that these two fixes did the trick.
